numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -2923,8 +2923,36 @@ static PyObject *
 
     /* TODO: include type name in error message, which is not @name@ */
     PyObject *obj = NULL;
+#if defined(_@TYPE@_IS_CDOUBLE) || defined(_@TYPE@_IS_CFLOAT)
+    static char *kwnames[] = {"", "", NULL};  /* positional-only */
+    static char *parse_error = "Could not convert arguments into a complex scalar. '%R' given.";
+    PyObject *real_obj = NULL;
+    PyObject *imag_obj = NULL;
+    if (PyArg_ParseTupleAndKeywords(args, kwds, "|OO", kwnames, &real_obj, &imag_obj)) {
+        if (imag_obj == NULL) {
+            obj = real_obj;
+            Py_XINCREF(obj);
```

Comment:
The refcounting for this is off below, hard to fix without a `goto finish` or so probably.  Might be easier to just include the scalar creation/assignment in the branch directly (duplicating some code)...

Reply:
Can you explain a bit more what the refcounting issue is? I think with the latest push everything is correct. 

We `INCREF` `obj` if it gets assigned to `real_obj`, but then a decref happens after `obj` gets used in `PyArray_FromAny`. There's an early return before that in the `obj == NULL` case, but then in that case we can't have any refcounting errors because `obj` is `NULL`.

There was a refcounting issue in the original version of this PR if getting the typecode failed, but as you pointed out in another comment, that's impossible given the way it's written. I deleted the error case and left a comment that it can't fail to make that clearer.

Reply:
Ah, yes you are right, I thought there were more error paths, but there is mainly that `obj == NULL` block.